Form layout—Documentation (HDS-4886) (#2947)
hashicorp/design-system: website/docs/patterns/form-patterns/partials/guidelines/guidelines.md

## Spacing

Spacing between sections, fields, text, and other form elements should follow a decreasing scale on an 8px grid. This creates a consistent visual hierarchy between elements.

### Form

> A **form component** acts as a layout mechanism by wrapping the content, fields, and actions into sections with consistent spacing while simultaneously handling the logic for submitting information and data collected by the fields.

Reviewer: ```suggestion
A form acts as a layout mechanism by wrapping the content, fields, and actions into sections with consistent spacing while simultaneously handling the logic for submitting information and data collected by the fields.
```

Author: It seems kind of strange to me to talk about the Form component behavior like this in the context of the form patterns. Maybe say something simpler in this context like:

```
A form acts as a container for form sections, fields, and other form content.
```

Or perhaps delete the whole "Form" section here entirely.

What do you think? @heatherlarsen

Reviewer: +1. This was in the original content, but "component" got added, so I suggested removing that, as I agree that it's strange since it's not a component, but an HTML element. I'd love to see the overall section simplified, and I like your suggestion.

Author: Thanks, I will update

### Sections

> A **section** occupies the largest hierarchy within the form. Sections organize text, fields (inputs, checkboxes, toggles, etc.), and actions into logical groups. A form consisting of more than one section should use a 32px gap between sections.

Reviewer: Removing "actions" because those are handled in the footer.

```suggestion
A section occupies the largest hierarchy within the form. Sections organize text and fields (inputs, checkboxes, toggles, etc.) into logical groups. A form consisting of more than one section should use a 32px gap between sections.
```

| [Toggle](/components/form/toggle) | Lorem ipsum | Lorem ipsum |
-->

### Field group

> If fields are collecting related information or information that is part of the same object, they may be organized in a field group. Within a field group there should be a 16px gap either horizontally, or vertically between each field.

Reviewer: ```suggestion
If fields are collecting related information or information that is part of the same object, they may be organized in a field group. Field groups are often organized in rows with a 16px gap between fields.
```

![Example of fields organized in a set](/assets/patterns/form-patterns/field-set.png =450x*)

Common sets of field groups can include:

- Credit card information: card number, expiration date, security code
- Multi-line address field
- First and last name

#### Alignment

Fields organized in a horizontal field group should be aligned to the baseline of each element to account for fields with helper text.

![Baseline alignment within a set](/assets/patterns/form-patterns/baseline-alignment.png =500x*)

### Button Sets

Organize buttons based on the [Button Set](/components/button-set) guidelines, e.g., using a 16px horizontal gap between buttons. For more specifications and examples, refer to the documentation for [button organization](/patterns/button-organization).

![Button set](/assets/patterns/form-patterns/button-set.png =500x*)

## Layout

### Single-column

> We recommend using a single-column layout for forms to streamline data entry. Forms across our products should use a maximum width; this can be achieved by using the Form Layout component or by using a size or unit relative to the page, viewport, or container size, e.g., viewport width unit (`vw`), percentage width (`50%`), or a character unit (`ch`).

Reviewer: I think this is the correct URL, but please fix, if not.

```suggestion
We recommend using a single-column layout for forms to streamline data entry. Forms across our products should use a maximum width which can be achieved by using the [Form Layout component](/components/form/layout) or by using a size or unit relative to the page, viewport, or container size, e.g., viewport width unit (`vw`), percentage width (`50%`), or a character unit (`ch`).
```

### Multi-column

> Forms follow a [single column layout](/components/form/layout#single-column-layout). However, you can use the [Section Multi Field Group](/components/form/layout#form-section-multi-field-group) to place multiple fields in a row when appropriate.

Reviewer: ```suggestion
While forms should typically follow a [single column layout](/components/form/layout#single-column-layout), you can use the [Section Multi Field Group](/components/form/layout#form-section-multi-field-group) to place multiple fields in a row, as needed.
```

| [Toggle](/components/form/toggle) | Lorem ipsum | Lorem ipsum |
-->

> ### Field group

Author: @LilithJames-HDS What do you think?

If fields are collecting related information or information that is part of the same object, they may be organized in a field group. Within a field group there should be a 16px gap either horizontally, or vertically between each field.

![Example of fields organized in a set](/assets/patterns/form-patterns/field-set.png =450x*)

Common sets of field groups can include:

- Credit card information: card number, expiration date, security code
- Multi-line address field
- First and last name

#### Alignment

Fields organized in a horizontal field group should be aligned to the baseline of each element to account for fields with helper text.

![Baseline alignment within a set](/assets/patterns/form-patterns/baseline-alignment.png =500x*)

### Button Sets

Organize buttons based on the [Button Set](/components/button-set) guidelines, e.g., using a 16px horizontal gap between buttons. For more specifications and examples, refer to the documentation for [button organization](/patterns/button-organization).

![Button set](/assets/patterns/form-patterns/button-set.png =500x*)

## Layout

### Single-column

> We recommend using a single-column layout for forms to streamline data entry. Since a relative width will adjust based on the viewport and container width, we recommend using a maximum and minimum width to ensure a smooth transition at any size.

Reviewer 2: > we recommend using a maximum and minimum width to ensure a smooth transition at any size

As a developer, I am not even sure how can I do it (not even with the new layout components)

Forms across our products should use a maximum width; this can be achieved by using the Form Layout component or by using a size or unit relative to the page, viewport, or container size, e.g., viewport width unit (`vw`), percentage width (`50%`), or a character unit (`ch`).

### Multi-column

Field groups create the appearance of a multi-column layout within the form.

> !!! Do

Author: @LilithJames-HDS Could you please take a look at this part?

!!! Dont

> Don't exceed more than three fields in a horizontal set. In most scenarios, it's best to limit the number of columns in a form or section to two.

Reviewer: ```suggestion
Don't exceed more than three fields in a field group. In most scenarios, it's best to limit the number of columns in a form or section to two.
```

![Three field maximum in a set](/assets/patterns/form-patterns/multi-column-three-fields-max.png =500x*)
!!!

!!! Do

If using multiple inputs in a row, stack fields vertically at the `md` breakpoint. This behavior is automatically applied in the Form Layout.

![Stack fields](/assets/patterns/form-patterns/multi-column-stacking.png =350x*)
!!!

## Width

The width of a form and the fields it contains are largely dependent on the form’s context. Adhering to these high-level guidelines can make a complex form approachable and introduce consistency in the UX application.

### Field width

Generally, a field’s width should account for the estimated width of the content it accepts. This gives the user an accurate sense of the character length and type of content the field accepts, which is important in setting user expectations.

### Responsive properties

> As the viewport shrinks, the form should expand relative to the viewport width, eventually occupying the entire width of the content area. Horizontal field groups should stack vertically as the viewport shrinks.

Reviewer: ```suggestion
As the viewport shrinks, the form should expand relative to the viewport width, eventually occupying the entire width of the content area. Field groups should stack vertically as the viewport shrinks.
```

!!! Dont

Don't exceed more than three fields in a horizontal set. In most scenarios, it's best to limit the number of columns in a form or section to two.

![Three field maximum in a set](/assets/patterns/form-patterns/multi-column-three-fields-max.png =500x*)
!!!

!!! Do

If using multiple inputs in a row, stack fields vertically at the `md` breakpoint. This behavior is automatically applied in the Form Layout.

> ![Stack fields](/assets/patterns/form-patterns/multi-column-stacking.png =350x*)

Author: @LilithJames-HDS Could you please check the image?
